src/Raven.Server/Integrations/PostgreSQL/RqlQuery.cs

```diff
@@ -175,90 +188,103 @@ public static bool TryParse(string queryText, int[] parametersDataTypes, Documen
 
         public override async Task Execute(MessageBuilder builder, PipeWriter writer, CancellationToken token)
         {
-            if (IsEmptyQuery)
-            {
-                await writer.WriteAsync(builder.EmptyQueryResponse(), token);
-                return;
-            }
-
-            if (_result == null)
-                throw new InvalidOperationException("RqlQuery.Execute was called when _results = null");
-
-            if (_limit == 0 || _result == null || _result.Count == 0)
-            {
-                await writer.WriteAsync(builder.CommandComplete($"SELECT 0"), token);
-                return;
-            }
-
-            BlittableJsonReaderObject.PropertyDetails prop = default;
-            var row = ArrayPool<ReadOnlyMemory<byte>?>.Shared.Rent(Columns.Count);
-
             try
             {
-                short? idIndex = null;
-                if (Columns.TryGetValue(Constants.Documents.Indexing.Fields.DocumentIdFieldName, out var col))
+                if (IsEmptyQuery)
                 {
-                    idIndex = col.ColumnIndex;
+                    await writer.WriteAsync(builder.EmptyQueryResponse(), token);
+                    return;
                 }
 
-                var jsonIndex = Columns[Constants.Documents.Querying.Fields.PowerBIJsonFieldName].ColumnIndex;
+                if (_result == null)
+                {
+                    if (IsNamedStatement)
+                        _result = await RunRqlQuery();
+                    else
+                        throw new InvalidOperationException("RqlQuery.Execute was called when _results = null");
+                }
 
-                foreach (var result in _result)
+                if (_limit == 0 || _result == null || _result.Count == 0)
                 {
-                    var jsonResult = result.Data;
+                    await writer.WriteAsync(builder.CommandComplete($"SELECT 0"), token);
+                    return;
+                }
 
-                    Array.Clear(row, 0, row.Length);
+                BlittableJsonReaderObject.PropertyDetails prop = default;
+                var row = ArrayPool<ReadOnlyMemory<byte>?>.Shared.Rent(Columns.Count);
```

Comment:
where are we returning this? BTW, it should be ArrayPool<byte>.Shared and then we should wrap it up, alternative option is to write MemoryManager that returns you exactly what you want and that takes leverge from the byte ArrayPool

Reply:
I see that we are returning the row in the [`finally` ](https://github.com/ravendb/ravendb/pull/15261/files#diff-de31fa030e49b5d9a259169d4bf9aef834714a07e75b4bb76e66397e5be3ad20R276-R280). 
What did you mean by wrapping it up? It's not my code so I might be missing something in it.
We're renting from ArrayPool.Shared here and returning in the finally.

Reply:
Ok, I see it now, switched to split view. Thx.